### PR TITLE
Added routes and controllers for what’s on

### PR DESCRIPTION
**Pull request by gestchild**
Splitting #1993 into more manageable chunks (will ultimately need all merged into master together, in order not to break anything. Once each PR is individual approved I'll merge them back down the chain)

### CODE REVIEW

File: whats-on/app/controllers.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  restaurantPromo,
+  spiritBoothPromo
+} from '../../server/data/daily-promos';
+const {createPageConfig} = model;
+const {
+  getExhibitionAndEventPromos
+} = prismic;
+
+export async function renderWhatsOn(ctx, next) {
+  const { 'f[dates]': timeSpan } = ctx.query;
```
Comment (jamesgorrie) [inline whats-on/app/controllers.js]: Could we use `startDate` / `endDate` here?

Comment (jamesgorrie) [reply whats-on/app/controllers.js]: It might also simplify the controller logic

File: server/services/prismic.js [hunk 2]
```diff
@@ -259,23 +254,146 @@ function convertPrismicResultsToPaginatedResults(prismicResults: Object): (resul
...
+        }
+        if (!acc[year][month]) {
+          acc[year][month] = [];
+        }
+        acc[year][month].push(currEvent);
+        start.add(1, 'month');
+      }
+    }
+    return acc;
+  }, {});
```
Comment (jamesgorrie) [inline server/services/prismic.js]: A common way to do this is [`groupBy`](http://facebook.github.io/immutable-js/docs/#/List/groupBy)

File: whats-on/app/controllers.js (hunk 1 above)
Comment (jamesgorrie) [reply whats-on/app/controllers.js]: FYI: https://github.com/pegjs/pegjs

File: exhibitions/app/views/pages/exhibitions.njk [hunk 3]
```diff
@@ -26,47 +26,43 @@
...
-          {% endif %}
-          {% if paginatedExhibitions.currentPage === paginatedExhibitions.totalPages %}
-            <div class="{{ {s: 12, m: 12, l: 12, xl: 12} | gridClasses }} {{ {s:4} | spacingClasses({margin: ['top']}) }}"  style="text-align: right">
-              {% componentV2 'more-info-link', {name: 'Past exhibitions', url: '/exhibitions/past'} %}
-            </div>
-          {% endif %}
-      </div>
+  <div class="css-grid__container  {{ {s:2, m:2, l:2} | spacingClasses({padding: ['top']}) }} {{ {s:3, m:3, l:3} | spacingClasses({padding: ['bottom']}) }} ">
+    <div class="css-grid">
+        {% if paginatedExhibitions.pagination.pageCount > 1  %}
```
Comment (davidpmccormick) [inline exhibitions/app/views/pages/exhibitions.njk]: a bit less indent?

File: exhibitions/app/views/pages/exhibitions.njk [hunk 4]
```diff
@@ -26,47 +26,43 @@
...
+          <div class="{{ {s: 12, m: 12, l: 12, xl: 12} | cssGridClasses }}"  style="text-align: right">
```
Comment (davidpmccormick) [inline exhibitions/app/views/pages/exhibitions.njk]: does this style have to be inline?

File: exhibitions/app/views/pages/exhibitions.njk [hunk 5]
```diff
@@ -26,47 +26,43 @@
...
+            {% componentV2 'pagination', paginatedExhibitions.pagination %}
+          </div>
+        {% endif %}
+        {% if paginatedExhibitions.currentPage === paginatedExhibitions.totalPages %}
+          <div class="{{ {s: 12, m: 12, l: 12, xl: 12} | cssGridClasses }} {{ {s:4} | spacingClasses({margin: ['top']}) }}"  style="text-align: right">
```
Comment (davidpmccormick) [inline exhibitions/app/views/pages/exhibitions.njk]: same with this one?

File: server/services/prismic.js [hunk 6]
```diff
@@ -184,7 +189,7 @@ export async function getCuratedList(id: string) {
 }
 
 function createExhibitionPromos(allResults: Object): Array<ExhibitionPromo> {
-  const allPromos = allResults.results.map((e):ExhibitionPromo => {
+  return allResults.map((e):ExhibitionPromo => {
```
Comment (davidpmccormick) [inline server/services/prismic.js]: Wonder if we should implement this eslint rule for consistency? https://github.com/gajus/eslint-plugin-flowtype/blob/master/.README/rules/space-before-type-colon.md

File: server/services/prismic.js [hunk 7]
```diff
@@ -259,23 +254,146 @@ function convertPrismicResultsToPaginatedResults(prismicResults: Object): (resul
...
+    permanentExhibitionPromos,
+    temporaryExhibitionPromos,
+    eventPromos,
+    eventPromosGroupedByMonth,
+    monthControls
+  };
+}
+
+function getWeekendFromDate(today) {
+  const todayInteger = today.day(); // day() return Sun as 0, Sat as 6
```
Comment (davidpmccormick) [inline server/services/prismic.js]: would `return london(today).day(todayInteger ? -2 : 5);` work?

File: server/services/prismic.js (hunk 6 above)
Comment (jamesgorrie) [reply server/services/prismic.js]: Could whack it in another PR?

File: server/services/prismic.js (hunk 7 above)
Comment (gestchild) [reply server/services/prismic.js]: It would, but I find it easier to follow how it is

File: server/services/prismic.js (hunk 2 above)
Comment (gestchild) [reply server/services/prismic.js]: This function isn't actually grouping the promos as such, but duplicating them (if their start and end dates span multiple months/years) and then grouping them accordingly (but the year and month doesn't exist as properties of the events). I've renamed the function so it's clearer.

Comment (jamesgorrie) [reply server/services/prismic.js]: `spanPromoAcrossMonthsAndYears`?
It's still not a concept I'm grabbing intrinsically, but I think I am being dim.

Comment (jamesgorrie) [reply server/services/prismic.js]: Would a diagram similar to what you drew help (it's how I respond well to Rx stuff).
```
                [weekend]
[event1_____]     [event2_____]
``` 

Don't feel obliged, I can just stare at it longer to figure out, I think this is my issue.

Comment (gestchild) [reply server/services/prismic.js]: the diagram was for the datesOverlapRange function and yep I'll add that

**Pull request merged**